cache-utils.go

```diff
@@ -0,0 +1,140 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+)
+
+// ResponseRecorder returns a wrapped response writer to
+// get underlying http.Response for cache handler.
+type ResponseRecorder struct {
+	StatusCode   int
+	header       http.Header // header needed for Header() method
+	Flushed      bool        // Flushed is whether the Handler called Flush.
+	ch           chan bool   // blocks until header is written or Write started
+	sentResponse bool
+	wroteHeader  bool
+
+	pw     *io.PipeWriter
+	result *http.Response // http response
+}
+
+// NewRecorder returns an initialized ResponseRecorder.
+func NewRecorder() *ResponseRecorder {
+	pr, pw := io.Pipe()
+	nr := &ResponseRecorder{
+		ch:         make(chan bool),
+		header:     make(http.Header),
+		StatusCode: 200,
+		pw:         pw,
+		result: &http.Response{
```

Comment:
do you need this? perhaps you can just keep `pw, pr *io.PipeWriter, *io.PipeReader`

Reply:
result is needed in the cache handler - not following your suggestion here.